vendor/github.com/xtaci/smux/session.go

```diff
@@ -281,73 +276,76 @@ func (s *Session) recvLoop() {
 	}
 }
 
-func (s *Session) keepalive() {
-	tickerPing := time.NewTicker(s.config.KeepAliveInterval)
-	tickerTimeout := time.NewTicker(s.config.KeepAliveTimeout)
-	defer tickerPing.Stop()
-	defer tickerTimeout.Stop()
+// keepAliveSend will periodically send a keepalive meesage to the remote peer.
+func (s *Session) keepAliveSend() {
+	keepAliveTimeout := time.After(s.config.KeepAliveInterval)
 	for {
 		select {
-		case <-tickerPing.C:
-			s.writeFrame(newFrame(cmdNOP, 0))
-			s.notifyBucket() // force a signal to the recvLoop
-		case <-tickerTimeout.C:
-			if !atomic.CompareAndSwapInt32(&s.dataReady, 1, 0) {
-				s.Close()
-				return
-			}
 		case <-s.die:
 			return
+		case <-keepAliveTimeout:
+			s.writeFrame(newFrame(cmdNOP, 0), time.Now().Add(s.config.WriteTimeout))
+			s.notifyBucket() // force a signal to the recvLoop
+			keepAliveTimeout = time.After(s.config.KeepAliveInterval)
 		}
 	}
 }
 
-func (s *Session) sendLoop() {
-	buf := make([]byte, (1<<16)+headerSize)
+// keepAliveTimeout will periodically check that some sort of message has been
+// sent by the remote peer, closing the session if not.
+func (s *Session) keepAliveTimeout() {
+	timeoutChan := time.After(s.config.KeepAliveTimeout)
 	for {
 		select {
 		case <-s.die:
 			return
-		case request, ok := <-s.writes:
-			if !ok {
-				continue
-			}
-			buf[0] = request.frame.ver
-			buf[1] = request.frame.cmd
-			binary.LittleEndian.PutUint16(buf[2:], uint16(len(request.frame.data)))
-			binary.LittleEndian.PutUint32(buf[4:], request.frame.sid)
-			copy(buf[headerSize:], request.frame.data)
-			n, err := s.conn.Write(buf[:headerSize+len(request.frame.data)])
-
-			n -= headerSize
-			if n < 0 {
-				n = 0
-			}
-
-			result := writeResult{
-				n:   n,
-				err: err,
+		case <-timeoutChan:
+			if !atomic.CompareAndSwapInt32(&s.dataWasRead, 1, 0) {
+				s.Close()
+				return
 			}
-
-			request.result <- result
-			close(request.result)
+			timeoutChan = time.After(s.config.KeepAliveTimeout)
 		}
 	}
 }
 
 // writeFrame writes the frame to the underlying connection
 // and returns the number of bytes written if successful
-func (s *Session) writeFrame(f Frame) (n int, err error) {
-	req := writeRequest{
-		frame:  f,
-		result: make(chan writeResult, 1),
+func (s *Session) writeFrame(frame Frame, timeout time.Time) (int, error) {
```

Comment:
writing a frame used to streamline through a thread that was listening on a channel for frames to send. Working in that setup required a lot of extra structs and channels, and made it easy to make mistakes. So easy, in fact, that there were deadlocks in the original code, and timeouts were not handled correctly at all.

Now, timeouts are handled correctly, things are still streamlined with the use of mutexes, and we've done away with all the extra code required to make this work with channels.

This is definitely a case where a mutex makes more sense. We aren't trying to pass information between actors, we are trying to make sure that only one actor is performing an action at a time.

Reply:
streamline -> serialize :P 

I do agree that a channel is not optimal here. It would make sense if the frames needed to be ordered, but they don't. It was probably chosen to enable both serialization and timeouts. (`TryMutex` uses a channel internally for the same reason).
However, I'm also a little uneasy overhauling the concurrency of a package that we don't fully understand. We obviously need to test this code thoroughly on the real network before including it in a release. But I'm also curious whether we can fix the bug we're observing while making minimal changes to smux.

Reply:
There's some discussion of deadlines here: https://github.com/xtaci/smux/issues/3

I suggest that we try calling `SetWriteDeadline` throughout the smux code and see if it breaks anything.